staging/src/k8s.io/client-go/tools/cache/shared_informer_health.go

```diff
@@ -0,0 +1,182 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+func newDefaultListWatchLockFreeHealth(name string, errorReceiverCh <-chan error) *listWatchLockFreeHealth {
+	lwlf := &listWatchLockFreeHealth{
+		name:                name,
+		minTimeToFail:       time.Duration(5 * time.Second),
+		minConsistentErrors: 10,
+		consistentErrorsCh:  errorReceiverCh,
+		errCircularBuffer:   atomic.Value{},
+	}
+
+	lwlf.errCircularBuffer.Store(newErrBuffer(5))
+	return lwlf
+}
+
+type listWatchLockFreeHealth struct {
+	name         string
+	startingTime int64
+
+	// minTimeToFail minimum time duration after which we start to fail
+	minTimeToFail time.Duration
+	// minConsistentError minimum number of errors before we start to fail
+	minConsistentErrors int32
+	// consistentErrors holds the number of errors seen so far
+	consistentErrors int32
+
+	// consistentErrorsCh channel on which we receive List&Watch related errors
+	// it is meant to represent a gauge, that is a number that can go up and down
+	consistentErrorsCh <-chan error
+
+	errCircularBuffer atomic.Value
+}
+
+func (lwlf *listWatchLockFreeHealth) run(stopCh <-chan struct{}) {
+	go wait.Until(func() {
+		lwlf.runInternal(stopCh)
+	}, 1*time.Minute, stopCh)
+}
+
+func (lwlf *listWatchLockFreeHealth) runInternal(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		case errRcv, ok := <-lwlf.consistentErrorsCh:
+			if !ok {
+				break
+			}
+			seenErrors := atomic.LoadInt32(&lwlf.consistentErrors)
+			if errRcv == nil {
+				seenErrors = 0
+				errBuf := lwlf.errCircularBuffer.Load().(errBuffer)
+				zeroErrBuf := errBuf.zero()
+				lwlf.errCircularBuffer.Store(zeroErrBuf)
+			} else {
+				seenErrors = seenErrors + 1
+				errBuf := lwlf.errCircularBuffer.Load().(errBuffer)
+				errBufCpy := errBuf.copy()
+				errBufCpy.add(errRcv)
+				lwlf.errCircularBuffer.Store(errBufCpy)
+			}
+			localStartingTime := atomic.LoadInt64(&lwlf.startingTime)
+			atomic.StoreInt32(&lwlf.consistentErrors, seenErrors)
+			if seenErrors >= lwlf.minConsistentErrors && localStartingTime <= 0 {
+				atomic.StoreInt64(&lwlf.startingTime, time.Now().UnixNano())
+			} else if seenErrors < lwlf.minConsistentErrors && localStartingTime > 0 {
+				atomic.StoreInt64(&lwlf.startingTime, 0)
+			}
+		}
+	}
+}
+
+// check conforms to HealthzChecker interface and reports health of an informer,
+// it returns an error if it observed more than minConsistentError for more than minTimeToFail
+// returned error contains the total number of errors seen so far,
+// elapsed time since the point in time it received more than minConsistentErrors
+// and the last n (set to 5 by default) errors received so far.
+func (lwlf *listWatchLockFreeHealth) check(_ *http.Request) error {
```

Comment:
for now this method can be private, will be public once we change `SharedInformer ` https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L43